src/module/normalisation/heap_profiler.rs

```diff
@@ -0,0 +1,219 @@
+// optional last argument. This last argument is a return value from "name"
+// function if such exists.
+fn add_profiler_to(module: &mut Module, name: &str) -> Result<()> {
+    // find local function in module
+    let fn_id = module
+        .funcs
+        .by_name(name)
+        .ok_or(anyhow::Error::msg(format!(
+            "heap_profiler: '{}' was not found in wasm",
+            name
+        )))?;
+    let types = module.types.params_results(module.funcs.get(fn_id).ty());
+    let (params, results) = (types.0.to_vec(), types.1.to_vec());
+
+    // Import profiler. Profilers don't return anything. Profilers last argument
+    // is result from the original function. For example, local function "malloc(i32) -> u32"
+    // will import profiler of type "malloc(i32, u32)".
+    let profiler_type = module
+        .types
+        .add(&[params.clone(), results.clone()].concat(), &[]);
+    let profiler_id = module
+        .add_import_func(
+            "heap_profiler",
+            &format!("{}_profiler", name),
+            profiler_type,
+        )
+        .0;
+
+    // create a clone of a function
+    let fn_copy_id = clone_function(module, fn_id, Some(format!("{}_wrap", name)));
+
+    let locals = &mut module.locals;
+    // create new local params for wrapper function, old params are copied (see clone above) to new
+    // function
+    let local_vars: Vec<LocalId> = params.iter().map(|t| locals.add(*t)).collect();
+    let rets: Vec<LocalId> = results.iter().map(|t| locals.add(*t)).collect();
+    let mut instr_seq = module
+        .funcs
+        .get_mut(fn_id)
+        .kind
+        .unwrap_local_mut()
+        .builder_mut()
+        .func_body();
+
+    // remove all instructions from wrapper function (they are copied over to new function)
+    *instr_seq.instrs_mut() = vec![];
+
+    // prepare args to call new function
+    local_vars.iter().for_each(|l| {
+        instr_seq.local_get(*l);
+    });
+
+    // call new copied function from wrapper function
+    instr_seq.call(fn_copy_id);
+
+    // save returned values from the above function
+    rets.iter().for_each(|r| {
+        instr_seq.local_set(*r);
+    });
+
+    // prepare args to call profiler function
+    local_vars.iter().for_each(|l| {
+        instr_seq.local_get(*l);
+    });
+    rets.iter().for_each(|r| {
+        instr_seq.local_get(*r);
+    });
+
+    // call profiler function
+    instr_seq.call(profiler_id);
+
+    // return saved values from original function
+    rets.iter().for_each(|r| {
+        instr_seq.local_get(*r);
+    });
+
+    // modify wrapper function args
+    module.funcs.get_mut(fn_id).kind.unwrap_local_mut().args = local_vars;
+    Ok(())
+}
+
+// TODO: move to normalisation/utils.rs ?
+// Create a new local function that will have same signiture and same
+// instructions as the supplied local function.
+fn clone_function(module: &mut Module, fn_id: FunctionId, name: Option<String>) -> FunctionId {
```

Comment:
Cloning walrus local function is not that straightforward. This implements cloning of a function with all its parameters and instructions and can be reused elsewhere. If this is the case we could move these functions to `utils` module

Reply:
Can we contribute something to Walrus?

Reply:
I will open an issue there